### PR TITLE
Winsorize care home outliers

### PR DESCRIPTION
**Pull request by garypolzin**
# Description
Previously we nulled outliers but this job now winsorizes outliers at the min/max permitted ratios

# How to test
[Default branch run](https://eu-west-2.console.aws.amazon.com/states/home?region=eu-west-2#/v2/executions/details/arn:aws:states:eu-west-2:344210435447:execution:windsorize-ch-outliers-Ind-CQC-Filled-Post-Estimates-Pipeline:58a61293-dfc6-4476-bf47-88899a19b311)

[Trello](https://trello.com/c/Lphwopon/760-care-home-filtering-winsorizing-outliers)

# Developer checklist
- [X] Unit test
- [X] Linked to Trello ticket
- [X] Documentation up to date

### CODE REVIEW

File: utils/ind_cqc_filled_posts_utils/clean_ascwds_filled_post_outliers/winsorize_care_home_filled_posts_per_bed_ratio_outliers.py [hunk 1]
```diff
@@ -0,0 +1,413 @@
...
+
+    care_homes_df = duplicate_ratios_within_standardised_residual_cutoffs(care_homes_df)
+
+    care_homes_df = calculate_min_and_max_permitted_filled_posts_per_bed_ratios(
+        care_homes_df
+    )
+
+    winsorized_df = winsorize_outliers(care_homes_df)
+
+    # TODO: do we want to identify which values have been winsorized? how?
```
Comment (MarionJHolloway) [inline utils/ind_cqc_filled_posts_utils/clean_ascwds_filled_post_outliers/winsorize_care_home_filled_posts_per_bed_ratio_outliers.py:80]: Yes, I think so and the branch is currently failing on this. Can we just use the same conditions as in the function winsorise outliers?

File: utils/ind_cqc_filled_posts_utils/clean_ascwds_filled_post_outliers/winsorize_care_home_filled_posts_per_bed_ratio_outliers.py [hunk 2]
```diff
@@ -0,0 +1,413 @@
...
+        & (
+            F.col(IndCQC.ascwds_filled_posts_dedup_clean).isNotNull()
+            & (F.col(IndCQC.ascwds_filled_posts_dedup_clean) > 0.0)
+        )
+    )
+
+    return df
+
+
+def select_data_not_in_subset_df(
```
Comment (MarionJHolloway) [inline utils/ind_cqc_filled_posts_utils/clean_ascwds_filled_post_outliers/winsorize_care_home_filled_posts_per_bed_ratio_outliers.py:117]: Missing docstring

File: utils/ind_cqc_filled_posts_utils/clean_ascwds_filled_post_outliers/winsorize_care_home_filled_posts_per_bed_ratio_outliers.py [hunk 3]
```diff
@@ -0,0 +1,413 @@
...
+    complete_df: DataFrame, subset_df: DataFrame
+) -> DataFrame:
+    output_df = complete_df.exceptAll(subset_df)
+
+    return output_df
+
+
+def create_banded_bed_count_column(
```
Comment (MarionJHolloway) [inline utils/ind_cqc_filled_posts_utils/clean_ascwds_filled_post_outliers/winsorize_care_home_filled_posts_per_bed_ratio_outliers.py:125]: Missing doc string

File: utils/ind_cqc_filled_posts_utils/clean_ascwds_filled_post_outliers/winsorize_care_home_filled_posts_per_bed_ratio_outliers.py [hunk 4]
```diff
@@ -0,0 +1,413 @@
...
+    )
+
+    number_of_beds_with_bands_df = set_banded_boundaries.setHandleInvalid(
+        "keep"
+    ).transform(number_of_beds_df)
+
+    return input_df.join(number_of_beds_with_bands_df, IndCQC.number_of_beds, "left")
+
+
+def calculate_average_filled_posts_per_banded_bed_count(
```
Comment (MarionJHolloway) [inline utils/ind_cqc_filled_posts_utils/clean_ascwds_filled_post_outliers/winsorize_care_home_filled_posts_per_bed_ratio_outliers.py:143]: Missing doc string

File: utils/ind_cqc_filled_posts_utils/clean_ascwds_filled_post_outliers/winsorize_care_home_filled_posts_per_bed_ratio_outliers.py [hunk 5]
```diff
@@ -0,0 +1,413 @@
...
+    output_df = input_df.groupBy(F.col(IndCQC.number_of_beds_banded)).agg(
+        F.avg(IndCQC.filled_posts_per_bed_ratio).alias(
+            IndCQC.avg_filled_posts_per_bed_ratio
+        )
+    )
+
+    return output_df
+
+
+def calculate_expected_filled_posts_based_on_number_of_beds(
```
Comment (MarionJHolloway) [inline utils/ind_cqc_filled_posts_utils/clean_ascwds_filled_post_outliers/winsorize_care_home_filled_posts_per_bed_ratio_outliers.py:155]: Missing doc string

File: utils/ind_cqc_filled_posts_utils/clean_ascwds_filled_post_outliers/winsorize_care_home_filled_posts_per_bed_ratio_outliers.py [hunk 6]
```diff
@@ -0,0 +1,413 @@
...
+
+    aggregated_df = set_minimum_permitted_ratio(
+        aggregated_df,
+        IndCQC.min_filled_posts_per_bed_ratio,
+        NumericalValues.MINIMUM_PERMITTED_LOWER_RATIO_CUTOFF,
+    )
+    aggregated_df = set_minimum_permitted_ratio(
+        aggregated_df,
+        IndCQC.max_filled_posts_per_bed_ratio,
+        NumericalValues.MINIMUM_PERMITTED_UPPER_RATIO_CUTOFF,
```
Comment (MarionJHolloway) [inline utils/ind_cqc_filled_posts_utils/clean_ascwds_filled_post_outliers/winsorize_care_home_filled_posts_per_bed_ratio_outliers.py:308]: Is this not the Max permitted? Might just need to rename function and magic number here

File: utils/ind_cqc_filled_posts_utils/clean_ascwds_filled_post_outliers/winsorize_care_home_filled_posts_per_bed_ratio_outliers.py [hunk 7]
```diff
@@ -0,0 +1,413 @@
...
+    maximum permitted ratios, the ascwds_filled_posts_dedup_clean is recalculated by multiplying the min/max
+    permitted ratio by the number_of_beds. The filled_posts_per_bed_ratio is then winsorized.
+
+    Args:
+        df (DataFrame): Input DataFrame.
+
+    Returns:
+        DataFrame: DataFrame with outliers winsorized.
+    """
+    lower_bound_condition = F.col(IndCQC.filled_posts_per_bed_ratio) < F.col(
```
Comment (MarionJHolloway) [inline utils/ind_cqc_filled_posts_utils/clean_ascwds_filled_post_outliers/winsorize_care_home_filled_posts_per_bed_ratio_outliers.py:355]: If using these for filtering column, will need to abstract - could go in dataclass at the top for ease of ref?

**Review by MarionJHolloway — CHANGES_REQUESTED** [review]
This is really hard to review given the size of the PR. Can we split it into two? One for renaming (which is easy to approve) and then another one which will then make the changes you've made more obvious? 

I have made some comments on what I think is the main point of this PR, but I'm not confident on what I'm actually supposed to be checking her because it's flagging everything as new!

File: utils/ind_cqc_filled_posts_utils/clean_ascwds_filled_post_outliers/winsorize_care_home_filled_posts_per_bed_ratio_outliers.py (hunk 6 above)
Comment (garypolzin) [reply utils/ind_cqc_filled_posts_utils/clean_ascwds_filled_post_outliers/winsorize_care_home_filled_posts_per_bed_ratio_outliers.py:308]: it's the mimimum permitted upper cutoff

we set this at 5

if the top 2.5% cutoff is calculated at 3.5, we would replace 3.5 with 5
if the top 2.5% cutoff is calculated at 18, we would use 18 as it's above the minimum permitted 5

File: utils/ind_cqc_filled_posts_utils/clean_ascwds_filled_post_outliers/winsorize_care_home_filled_posts_per_bed_ratio_outliers.py (hunk 1 above)
Comment (garypolzin) [reply utils/ind_cqc_filled_posts_utils/clean_ascwds_filled_post_outliers/winsorize_care_home_filled_posts_per_bed_ratio_outliers.py:86]: created a trello ticket for this

https://trello.com/c/kp8RLn7V/768-identify-values-changed-by-filtering-winsorizing

File: tests/test_file_data.py [hunk 8]
```diff
@@ -2989,99 +2989,99 @@ class CleanAscwdsFilledPostOutliersData:
...
-        ("37", date(2023, 1, 1), "Y", PrimaryServiceType.care_home_only, 25, 37.0, 37.0, AscwdsFilteringRule.populated, 1.48),
-        ("38", date(2023, 1, 1), "Y", PrimaryServiceType.care_home_only, 25, 38.0, 38.0, AscwdsFilteringRule.populated, 1.52),
-        ("39", date(2023, 1, 1), "Y", PrimaryServiceType.care_home_only, 25, 39.0, 39.0, AscwdsFilteringRule.populated, 1.56),
-        ("40", date(2023, 1, 1), "Y", PrimaryServiceType.care_home_only, 25, 40.0, 40.0, AscwdsFilteringRule.populated, 1.60),
-        ("41", date(2023, 1, 1), "Y", PrimaryServiceType.care_home_only, 25, None, None, AscwdsFilteringRule.missing_data, None),
-        ("42", date(2023, 1, 1), "Y", PrimaryServiceType.care_home_only, None, 42.0, 42.0, AscwdsFilteringRule.populated, None),
-        ("43", date(2023, 1, 1), "N", PrimaryServiceType.non_residential, 25, 43.0, 43.0, AscwdsFilteringRule.populated, 0.92),
-        ("44", date(2023, 1, 1), "N", PrimaryServiceType.non_residential, None, 44.0, 44.0, AscwdsFilteringRule.populated, None),
+        ("01", date(2023, 1, 1), "Y", PrimaryServiceType.care_home_only, 50, 2.0, 2.0, 0.04),
```
Comment (MarionJHolloway) [inline tests/test_file_data.py:2992]: Can you leave the filtering rule column in, but not updated? I can pick up the ticket to fix this next. And maybe add a comment after the ones that should have changed?

File: utils/ind_cqc_filled_posts_utils/clean_ascwds_filled_post_outliers/winsorize_care_home_filled_posts_per_bed_ratio_outliers.py [hunk 9]
```diff
@@ -6,16 +6,23 @@
...
-    DECIMAL_PLACES_TO_ROUND_TO: int = 5
-    PERCENTAGE_OF_DATE_TO_REMOVE_AS_OUTLIERS: float = 0.05
+class SetValuesForWinsorization:
+    """
+    Set numerical values for winsorization process.
+
+    Attributes:
+        PERCENTAGE_OF_DATA_TO_REMOVE_AS_OUTLIERS (float): Sets the proportion of data to identify as outliers (where 0.05 is 5%, which would identify the top and bottom 2.5% as outliers)
+        MINIMUM_PERMITTED_LOWER_RATIO_CUTOFF (float): Sets the lowest minimum filled_posts_per_bed_ratio permitted
+        MINIMUM_PERMITTED_UPPER_RATIO_CUTOFF (float): Sets the lowest maximum filled_posts_per_bed_ratio permitted
```
Comment (MarionJHolloway) [inline utils/ind_cqc_filled_posts_utils/clean_ascwds_filled_post_outliers/winsorize_care_home_filled_posts_per_bed_ratio_outliers.py:20]: I'm not sure where the comment you made has gone, but could you add the explanation here as to why this shouldn't be the max?

**Review by MarionJHolloway — APPROVED** [review]
Just a couple of small tweaks - have approved so I don't slow you down. Looks good though!

File: utils/ind_cqc_filled_posts_utils/clean_ascwds_filled_post_outliers/winsorize_care_home_filled_posts_per_bed_ratio_outliers.py (hunk 9 above)
Comment (garypolzin) [reply utils/ind_cqc_filled_posts_utils/clean_ascwds_filled_post_outliers/winsorize_care_home_filled_posts_per_bed_ratio_outliers.py:20]: your previous comment was where this value is used (not here where it's defined) so it's still there but further down :)

**Pull request merged**